Expensify/App
# Background push notification processing

### Details
This pull request makes it so that push notifications received when the Android app is closed will update Onyx, within the bounds of certain [OS restrictions](https://developer.android.com/topic/performance/power/power-details). This is important to achieve a background data trickle, and update the mobile app with chat data as frequently as possible.

On Android, push notifications received when the app is closed will wake a headless JS process (that lives in the UrbanAirship RN sdk) and execute a JS callback. _However_, it _will not_ mount any RN components, so the push notification callbacks must be registered outside of any React lifecycle. Because the callback is dependent on the Onyx store, we must also initialize that outside of any React lifecycle. Fortunately, that works just fine for our application 😁 

### Fixed Issues
$ (partial) https://github.com/Expensify/Expensify/issues/158830

### Tests
This is a bit of a doozy to test. It will be native-only, with only regression testing on other platforms.

#### Get a prod build running – Android
1. Follow the instructions in [this SO](https://stackoverflow.com/questions/35935060/how-can-i-generate-an-apk-that-can-run-without-server-with-react-native) to generate a signing key and provide it to the app through `build.gradle`.
1. Open `android/app/src/main/assets/airshipconfig.properties`. Swap out the `productionAppKey` and `productionAppSecret` with the `developmentAppKey` and `developmentAppSecret`.

    ![image](https://user-images.githubusercontent.com/47436092/124335793-6008fb00-db50-11eb-89ab-76ffdcfebf5a.png)

1. Run `npm i`
1. Run `react-native start --reset-cache`. Kill this command once you see the cache has been cleared.
1. Run `cd android && ./gradlew clean`
1. Open your Android emulator and make sure to fully uninstall any Expensify.cash app on there.
1. Run `npm run android -- --variant=release`.
1. Once that runs, verify that you do not see any logs in the metro window that opens. Kill the metro window, and verify that you can still view the app. Close the app (hit the square button and swipe up), then reopen it. It should still work.

#### Get a prod build running – iOS
1. This requires a physical iPhone for testing.
1. Run through the steps in [this SO](https://stackoverflow.com/c/expensify/questions/6735) to get set up to run the app on a physical iPhone device.
1. Run `npm i -g ios-deploy`
1. Plug your device into your computer. Click "trust" when prompted.
1. Open XCode, go to the `Expensify.cash` project configurations, click `Sigining and Capabilities`, and click "Enable Automatic". This will make some temporary changes that _must not_ be committed, but for me are always necessary to get the app to launch on a physical device.
1. Finally, run `npm run ios -- --configuration Release --device "Your Device Name"`

#### Test push notifications (both platforms)
1. Tail your VM logs for `[Report] Handled event sent by Airship`. This will indicate that the push notification callback executed.    
1. Run `npm run web` (you pick)
1. Sign into account A on the mobile app. Sign into account B on the web app. These should be accounts on your VM, not prod accounts.
1. Send a message from account B to account A. It should arrive normally.
1. Send a reply from account A to account B. It should arrive normally.
1. As user A, open a _different chat_, and background the app.
1. Send a message from B to A. A push notification should appear, and a log line demonstrating that the callback occurred should also appear.
1. Put the emulator/device on airplane mode.
1. Double-tap the notification. It should open E.cash to the correct chat, and the message should be available.
1. Take the emulator/device off airplane mode.
1. (android only) Open the emulator settings, then go to `Security` -> `Screen Lock` -> `Swipe`.

    ![image](https://user-images.githubusercontent.com/47436092/124336065-75caf000-db51-11eb-9d82-441d3804b56f.png) 
    
1. Hit the power button on the emulator menu (or lock button on the iPhone) to lock the emulator/device:

    ![image](https://user-images.githubusercontent.com/47436092/124336078-854a3900-db51-11eb-8904-749ce8bb8815.png)

1. Send a message from B to A. A push notification should appear, and a log line demonstrating that the callback occurred should also appear.
1. Put the emulator/device on airplane mode.
1. Double-tap the notification. It should open E.cash to the correct chat, and the new message should be available.
1. Hit the square button and swipe up to close E.cash entirely.
1. Send a message from B to A. A push notification should appear, and a log line demonstrating that the callback occurred should also appear.
1. Put the emulator/device on airplane mode.
1. Double-tap the notification. It should open E.cash to the correct chat, and the new message should be available.

### QA Steps
1. Sign into account A on the Android app. Sign into account B on the web app.
1. Send a message from account B to account A. It should arrive normally.
1. Send a reply from account A to account B. It should arrive normally.
1. As user A, open a _different chat_, and background the app.
1. Send a message from B to A. A push notification should appear.
1. After about 10 seconds, put the emulator on airplane mode.
1. Double-tap the notification. It should open E.cash to the correct chat, and the message should be available.
1. Take the emulator off airplane mode.
1. Open the android settings, then go to `Security` -> `Screen Lock` -> `Swipe` (doesn't have to be swipe – just make sure you can see a lock screen on the device). 
1. Lock the Android device.
1. Send a message from B to A. A push notification should appear.
1. After about 10 seconds, put the emulator on airplane mode.
1. Double-tap the notification. It should open E.cash to the correct chat, and the new message should be available.
1. Hit the square button and swipe up to close E.cash entirely.
1. Send a message from B to A. A push notification should appear, and a log line demonstrating that the callback occurred should also appear.
1. After about 10 seconds, put the emulator on airplane mode.
1. Double-tap the notification. It should open E.cash to the correct chat, and the new message should be available.

### Tested On


### Screenshots
#### Web

#### Mobile Web


#### Desktop


#### iOS


#### Android
....Will need to gather screen recordings after the long weekend.
